pkg/controller/daemon/update.go

```diff
@@ -64,20 +85,275 @@ func (dsc *DaemonSetsController) rollingUpdate(ds *extensions.DaemonSet) error {
 			break
 		}
 		glog.V(4).Infof("Marking pod %s/%s for deletion", ds.Name, pod.Name)
-		podsToDelete = append(podsToDelete, pod.Name)
+		oldPodsToDelete = append(oldPodsToDelete, pod.Name)
 		numUnavailable++
 	}
-	errors := dsc.syncNodes(ds, podsToDelete, []string{})
-	return utilerrors.NewAggregate(errors)
+	return dsc.syncNodes(ds, oldPodsToDelete, []string{}, hash)
+}
+
+func (dsc *DaemonSetsController) constructHistory(ds *extensions.DaemonSet) (cur *apps.ControllerRevision, old []*apps.ControllerRevision, err error) {
+	var histories []*apps.ControllerRevision
+	var currentHistories []*apps.ControllerRevision
+	histories, err = dsc.controlledHistories(ds)
+	if err != nil {
+		return
+	}
+	for _, history := range histories {
+		// Add the unique label if it's not already added to the history
+		// We use history name instead of computing hash, so that we don't need to worry about hash collision
+		if _, ok := history.Labels[extensions.DefaultDaemonSetUniqueLabelKey]; !ok {
```

Comment:
The `history` here are only those which are selected and controlled (owner ref) by the DS. Adding the label to the history here because current (new) history needs the unique label so that DS can label its pods.